libraries/standard/mqtt/src/iot_mqtt_api.c

```diff
@@ -726,6 +802,46 @@ static IotMqttError_t _subscriptionCreateAndSerialize( IotMqttOperationType_t op
 
 /*-----------------------------------------------------------*/
 
+static IotMqttError_t _sendSubscribeRequest( IotMqttOperationType_t operation,
+                                             IotMqttConnection_t mqttConnection,
+                                             _mqttOperation_t * pSubscriptionOperation,
+                                             uint32_t flags )
+{
+    IotMqttError_t status = IOT_MQTT_SUCCESS;
+
+    IotMqtt_Assert( pSubscriptionOperation != NULL );
+
+    /* Send the SUBSCRIBE packet. */
+    if( ( flags & MQTT_INTERNAL_FLAG_BLOCK_ON_SEND ) == MQTT_INTERNAL_FLAG_BLOCK_ON_SEND )
+    {
+        _IotMqtt_ProcessSend( IOT_SYSTEM_TASKPOOL, pSubscriptionOperation->job, pSubscriptionOperation );
+    }
+    else
+    {
+        status = _IotMqtt_ScheduleOperation( pSubscriptionOperation,
+                                             _IotMqtt_ProcessSend,
+                                             0 );
+
+        if( status != IOT_MQTT_SUCCESS )
+        {
+            IotLogError( "(MQTT connection %p) Failed to schedule %s for sending.",
+                         mqttConnection,
+                         IotMqtt_OperationType( operation ) );
+
+            if( operation == IOT_MQTT_SUBSCRIBE )
+            {
+                _IotMqtt_RemoveSubscriptionByPacket( mqttConnection,
+                                                     pSubscriptionOperation->u.operation.packetIdentifier,
+                                                     MQTT_REMOVE_ALL_SUBSCRIPTIONS );
+            }
+        }
```

Comment:
Except for this block, this function looks identical to `_sendPublish`. Perhaps they could be combined into one function and this logic could be moved back to `_subscriptionCommon`?

Reply:
_IotMqtt_ProcessSend() is common for all send functions, so it is already factored out. I'm not sure if complexity of _subscriptionCommon() will increase by doing that. I'll leave it that way right now, we can revisit later if needed.

Reply:
I have updated the function and and the complexity is still below 9,  so we are good.